compiler-core/src/language_server/type_annotations.rs

```diff
@@ -0,0 +1,204 @@
+use super::code_action::CodeActionBuilder;
+use crate::{
+    ast::{Statement, TypeAst, TypedFunction, TypedModuleConstant},
+    line_numbers::LineNumbers,
+    type_::{lsp_pretty::LspPrinter, pretty::Printer, Type, TypeVar},
+};
+use ecow::EcoString;
+use im::HashMap;
+use lsp_types::{
+    CodeActionKind, InlayHint, InlayHintKind, InlayHintLabel, Position, Range, TextEdit, Url,
+};
+
+#[derive(Debug, Clone, Default)]
+pub struct TypeAnnotations {
+    annotations: Vec<(Position, String)>,
+}
+
+impl TypeAnnotations {
+    pub fn from_module_constant(
+        constant: &TypedModuleConstant,
+        line_numbers: &LineNumbers,
+    ) -> Self {
+        if constant.annotation.is_none() {
+            let linecol = line_numbers.line_and_column_number(constant.location.end);
+            let position = Position::new(linecol.line - 1, linecol.column - 1);
+            let type_text = Printer::new().pretty_print(&constant.type_, 0);
+            let text = format!(": {type_text}");
+
+            Self {
+                annotations: vec![(position, text)],
+            }
+        } else {
+            Self::default()
+        }
+    }
+
+    pub fn extract_type_parameters_from_fn(function: &TypedFunction) -> HashMap<u64, TypeAst> {
```

Comment:
Why do we do this?

Reply:
Oh we do this to make a note of all user given type-parameter names anywhere in the function. We call this function before we process a function definition to generate inlay-hints. 
eg:
 `
fn equals(a: value, b) {
    a == b
} // In this case b's type would also be annotated as (value)
`